klippy/extras/firmware_retraction.py

```diff
@@ -1,74 +1,543 @@
+        # Zero min. and standard val to ensure compatibility with macros
+        self.z_hop_height = self.config_ref.getfloat(\
+            'z_hop_height', 0., minval=0.)
+        self.z_hop_style = self.config_ref.get(\
+            'z_hop_style', default='standard').strip().lower()
+        self._check_z_hop_style()   # Safe guard that zhop style is properly set
+        # verbose to enable/disable user messages
+        self.verbose = self.config_ref.getboolean('verbose', False)
+        # Control retraction parameter behaviour when retraction is cleared.
+        # Default is to reset retraction parameters to config values.
+        self.config_params_on_clear = self.config_ref.getboolean(\
+            'config_params_on_clear', True)
+
+    ################### Helper to check that z_hop_style was input and is valid.
+    def _check_z_hop_style(self):
+        if self.z_hop_style not in self.valid_z_hop_styles:
+            self.z_hop_style = 'standard'
+            logging.warning('The provided z_hop_style value is invalid. Using \
+                "standard" as default.')
+
+    ######## Helper method to register commands and instantiate required objects
+    def _handle_ready(self):
+        self.gcode = self.printer.lookup_object('gcode')
+        self.gcode_move = self.printer.lookup_object('gcode_move')
+        self.toolhead = self.printer.lookup_object('toolhead')
+        self.extruder = self.printer.lookup_object('extruder')
+
+        # Register new G-code commands for setting/retrieving retraction
+        # parameters and clearing retraction
+        self.gcode.register_command('SET_RETRACTION', self.cmd_SET_RETRACTION, \
+            desc=self.cmd_SET_RETRACTION_help)
+        self.gcode.register_command('GET_RETRACTION', self.cmd_GET_RETRACTION, \
+            desc=self.cmd_GET_RETRACTION_help)
+        self.gcode.register_command('CLEAR_RETRACTION', \
+            self.cmd_CLEAR_RETRACTION, desc=self.cmd_CLEAR_RETRACTION_help)
+
+        # Register new G-code commands for firmware retraction/unretraction
+        self.gcode.register_command('G10', self.cmd_G10)
+        self.gcode.register_command('G11', self.cmd_G11)
+        # Add M103 and M101 aliases for G10 and G11
+        self.gcode.register_command('M103', self.cmd_G10)
+        self.gcode.register_command('M101', self.cmd_G11)
+
+        # Register Events to clear retraction when a new print is started, an
+        # ongoing print is canceled or a print is finished
+        # Consider two operational modes: Printing from Virtual SD Card or via
+        # GCode streaming
+        ################ GCode streaming mode (most commonly done via OctoPrint)
+        # Print is started:  Most start gcodes include a G28 command to home all
+        # axes, which is generally NOT repeated during printing. Using homing as
+        # an indicator to evaluate if a printjob has started. G28 requirement
+        # added in function description.
```

Comment:
Is this the intended way to detect a started print?
I have some scripts that avoid homing if the printer already is homed. And this is happening in some situations.
If this is the intended way: Good to know, I have to change my scripts :)
If not this should possibly be analyzed to avoid some edge cases if retracted == true on auto bed leveling. Depending on implementation this could cause damage on the printer bed or nozzle.

Reply:
If this is the intended way and some edge cases could exist, a user recommendation to add clear_retraction into the beginning of the start gcode could make the user more sensitive against this issue.
I did read the text you wrote about this and I understand you did put some thought on this. However, it can be hard to respect or test all edge cases.
In the end this feature is worth a few broken beds until the bug reports are incoming, as it is really nice feature, and broken beds just are a given risk for 3D printing people ^^

Reply:
Hi masterxq! In case prints are done via gcode streaming (like OctoPrint does), it is almost impossible to figure out when a print is paused or finished. In the cases of start and canceled, I figured homing and stepper disable are good indicators or what is going on. You do indeed have a point with the bed mesh case in retract state. I checked it and changed the monitored event from home_rails_begin to homing_move_begin. Now this case also clears retraction when a bed_mesh_calibrate is called while retracted and already homed.

In any case, in the description I did add that START_PRINT macros should include a CLEAR_RETRACTION command. The homing move and stepper disable events are basically the firmware built-in redundancy in case the user forgets to include this command in the START_PRINT. If printing via virtualSDCard, print start is detected via events from the print_stats module. Hence, theoretically there is no need to add CLEAR_RETRACTION in the START_PRINT macro if the user always prints via virtual_sdcard (which is recommended also).

In terms of broken beds and nozzles, to prevent any damage in a super-edge case, I added a method which checks the z height after unretracting. If the result is that the nozzle may crash into the bed, the unretract zhop height is reduced accordingly. The code does not use gcode_offsets but instead leaves the printers coordinate systems untouched and changes the z-coordinate of move commands when in retract state. This is another safety feature.

Finally, I added a G1 command toggle flag just in case some weird edge case would meddle with the G1 command toggling (not that I would know of such an edge case but better safe than sorry). In summary, I believe that I built tons of redundancy into this code to make sure that it covers 99+% (maybe even all) edge cases.

Finally, thanks for your review, like and positive comments regarding this feature :)